demos/unsupported/transfer/transfer.man
.\" $XConsortium: transfer.man /main/4 1995/07/17 10:49:13 drk $
.\" Motif
.\"
.\" Copyright (c) 1987-2012, The Open Group. All rights reserved.
.\"
.\" These libraries and programs are free software; you can
.\" redistribute them and/or modify them under the terms of the GNU
.\" Lesser General Public License as published by the Free Software
.\" Foundation; either version 2.1 of the License, or (at your option)
.\" any later version.
.\"
.\" These libraries and programs are distributed in the hope that
.\" they will be useful, but WITHOUT ANY WARRANTY; without even the
.\" implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
.\" PURPOSE. See the GNU Lesser General Public License for more
.\" details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with these librararies and programs; if not, write
.\" to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
.\" Floor, Boston, MA 02110-1301 USA
...\" 
...\" 
...\" HISTORY
.TH transfer 1X MOTIF "Demonstration programs"
.SH NAME
\*Ltransfer\*O\(emUTM demonstration and utility
.SH SYNOPSIS
.sS
\*Ltransfer\*O
.sE
.SH DESCRIPTION
\*Ltransfer\*O
shows off the features in UTM and can aid in debugging data 
interchange problems.
.SH X Defaults
The class of this application is \fIXmDTransfer\fR.

To set a specific resource in your \fI.Xdefaults\fR file,  you specify the
particular widget and resource.


.attic/ChangeLog
2017-11-10 11:49:10 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1665 (HEAD, origin/master, origin/HEAD, master)

2017-10-30 23:04:15 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1651

2017-10-30 22:22:31 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1650

2017-10-30 21:47:21 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1652

2017-08-31 06:28:27 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1655

2017-08-28 05:33:53 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1625

2017-08-26 06:12:14 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1656

2017-08-23 04:10:55 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1658

2017-08-22 06:05:00 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1659

2017-08-22 03:27:36 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1660

2017-03-28 01:40:33 +0300 Oleksiy Chernyavskyy 

	* updated version and release data

2017-03-28 01:32:30 +0300 Oleksiy Chernyavskyy 

	* removed redundant includes

2017-03-28 01:11:59 +0300 Oleksiy Chernyavskyy 

	* removed bugfix 1565. See details on http://bugs.motifzone.net/show_bug.cgi?id=1565

2016-11-09 07:19:53 +0200 Alex 

	* obsolete malloc.h changed on stdlib.h

2016-06-11 05:03:21 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-11 05:02:55 +0300 Oleksiy Chernyavskyy 

	* updated release 2.3.6 information

2016-06-11 04:57:54 +0300 Oleksiy Chernyavskyy 

	* fixid parallel building (make -j) error

2016-06-11 04:57:20 +0300 Oleksiy Chernyavskyy 

	* removed automatic configure start from autogen.sh

2016-06-11 04:26:48 +0300 Oleksiy Chernyavskyy 

	* minor fix

2016-06-11 04:21:53 +0300 Oleksiy Chernyavskyy 

	* removed extra dist Imakefile from all Makefile.am files

2016-06-11 04:06:37 +0300 Oleksiy Chernyavskyy 

	* Purified copyright headers (Motif is LGPL covered). Removed many Imakefile's and Imake-related stuff

2016-06-10 23:40:43 +0300 Oleksiy Chernyavskyy 

	* updated release notes and version 2.3.5 -> 2.3.6

2016-06-10 23:27:56 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-10 23:15:00 +0300 Oleksiy Chernyavskyy 

	* bug 1654 fix

2016-06-10 21:58:16 +0300 Oleksiy Chernyavskyy 

	* old patches. need review

2016-03-29 03:48:59 +0300 Oleksiy Chernyavskyy 

	* removed obsolete files

2016-03-17 04:23:17 +0200 Oleksiy Chernyavskyy 

	* release 2.3.5 (tag: release_2_3_5)

2016-03-16 23:07:12 +0200 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-03-16 23:06:24 +0200 Oleksiy Chernyavskyy 

	* added mkchangelog script for automated generation of the ChangeLog

2016-03-16 00:46:49 +0200 Oleksiy Chernyavskyy 

	* Reimplemented bugfix 1565

2016-02-26 03:36:51 +0200 Oleksiy Chernyavskyy 

	* Updated RPM creation instructions

2014-08-20 15:45:41 +0300 Oleksiy Chernyavskyy 

	* fixed bug 1636

2013-08-21 11:49:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1612 (Label size computed wrong within a Form).

2013-07-11 18:14:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1618 (Eliminates 29 warnings of some type in modules from "demos" catalog).

2013-07-11 15:20:54 +0300 Mykola Vshyvkov 

	* Fixed bug #1616 (Eliminates 23 warnings "suggest parentheses around assignment used" in modules from "demos" catalog).

2013-07-11 15:01:50 +0300 Mykola Vshyvkov 

	* Fixed bug #1617 (Eliminates 23 warnings "cast to pointer from integer of different size" in modules from "demos" catalog).

2013-07-11 12:35:30 +0300 Mykola Vshyvkov 

	* Fixed bug #1615 (Eliminates 35 warnings "enumeration value XXX not handled" in modules from "demos" catalog).

2013-07-10 06:55:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1614 (Eliminates 36 warnings "variable may be used uninitialized" in modules from "demos" catalog).

2013-07-10 05:05:28 +0300 Mykola Vshyvkov 

	* Fixed bug #1613 (Eliminates 60 warnings "implicit declaration of function" in modules from "demos" catalog).

2013-07-04 14:08:09 +0300 Mykola Vshyvkov 

	* Fixed bug #1611 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls in modules of "clients" catalog).

2013-07-03 21:27:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1610 (Eliminates 15 warnings of some types in modules of "clients" catalog).

2013-07-03 18:52:21 +0300 Mykola Vshyvkov 

	* Fixed bug #1609 (Eliminates 18 warnings  of "incompatible implicit declaration" in modules of "clients" catalog).

2013-07-03 14:58:22 +0300 Mykola Vshyvkov 

	* Fixed bug #1608 (Eliminates 30 warnings  of "array subscript has type 'char'" in modules of "clients" catalog).

2013-07-03 14:35:04 +0300 Mykola Vshyvkov 

	* Fixed bug #1607 (Eliminates 92 warnings  of "variable may be used uninitialized" in modules of "clients" catalog).

2013-07-03 12:48:55 +0300 Mykola Vshyvkov 

	* Fixed bug #1606 (Eliminates 130 warnings  of "expects argument of type" in modules of "clients" catalog).

2013-07-02 13:40:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 20:52:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 19:07:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1604 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls).

2013-07-01 19:02:25 +0300 Mykola Vshyvkov 

	* Fixed bug #1605 (Eliminates 14 warnings of some types in WML tool).

2013-06-26 23:45:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1603 (Eliminates 18 warnings of some types in some modules).

2013-06-26 19:21:20 +0300 Mykola Vshyvkov 

	* Fixed bug #1602 (Eliminates 6 warnings "implicit declaration of function" in some modules).

2013-06-26 18:36:47 +0300 Mykola Vshyvkov 

	* Fixed bug #1601 (Eliminates 6 warnings: "suggest explicit braces to avoid ambiguous" in some modules).

2013-06-26 17:40:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1600 (Eliminates 26 warnings about using uninitialized variables in some modules).

2013-06-26 15:34:53 +0300 Mykola Vshyvkov 

	* Fixed bug #1599 (Eliminates 14 warnings of incompatible type argument from some modules).

2013-06-25 17:37:02 +0300 Mykola Vshyvkov 

	* Updated ChangeLog.

2013-06-25 13:57:31 +0300 Mykola Vshyvkov 

	* Merge branch 'openmotif_2_3' of ssh://git.code.sf.net/p/motif/code into openmotif_2_3


demos/unsupported/transfer/transfer.c
/* 
 * Motif
 *
 * Copyright (c) 1987-2012, The Open Group. All rights reserved.
 *
 * These libraries and programs are free software; you can
 * redistribute them and/or modify them under the terms of the GNU
 * Lesser General Public License as published by the Free Software
 * Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * These libraries and programs are distributed in the hope that
 * they will be useful, but WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with these librararies and programs; if not, write
 * to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301 USA
 */ 
/* 
 * HISTORY
 */ 

#ifdef REV_INFO
#ifndef lint
static char rcsid[] = "$XConsortium: transfer.c /main/6 1995/07/14 09:47:23 drk $"
#endif
#endif

#include <stdio.h>
#include <Xm/Xm.h>           /* Motif Toolkit */
#include "transfer.h"

void quitCB(Widget widget, char *tag, XmAnyCallbackStruct *callback_data);
void setMode(Widget widget, int param, XmAnyCallbackStruct *callback_data);
void cut(Widget widget, char *ignore, XmAnyCallbackStruct *callback_data);
void copy(Widget widget, char *ignore, XmAnyCallbackStruct *callback_data);
void paste(Widget widget, char *ignore, XmAnyCallbackStruct *callback_data);
void pasteMove(Widget widget, char *ignore, 
	       XmAnyCallbackStruct *callback_data);

static void NOP(Widget widget, XEvent *event, 
		String *params, Cardinal *num_params);
static int ErrorHandler(Display *display, XErrorEvent *event);

/* String to use for Clipboard tagging */
XmString generic_label;
Display  	*display;
XtAppContext  	app_context;

/* This tracks which transfer selection we're using.  Drag and drop isn't
   listed since it is dynamic,  we don't take ownership of DnD */
TransferMode Tmode = Mprimary;

/* We need a NOP action */
static XtActionsRec actions[] = {
  { "NOP",		NOP},
};

/*************************************************************************
 * A list of widgets we need to be able to find outside the main
 * program.
 *
 * toplevel	is the ApplicationShell
 * mainw	is the child of toplevel,  an XmMainWindow
 * target	is the drawing area used for drag and drop
 * list		is the list of targets displayed
 * text1	contains the value of the selected target
 * text2	contains the name of the selected or new target
 * typeOM	is the option menu which displays type information for the
 *  		selected or new target
 * format_om    Format option menu
 * format_s	String format button
 * format_l	Locale format button
 * format_t	CompoundText format button
 * format_cs	CompoundString format button
 * format_id	Integer format button (decimal)
 * format_hd	Integer format button (hex)
 *************************************************************************/

Widget toplevel, mainW, editBox;

/*
 *  Main program
 */

int 
main(int argc, String *argv)
{
    Arg	args[10];
    int n;

    toplevel = XtVaAppInitialize(&app_context, "XmdTransfer", NULL, 0, 
				 &argc, argv, NULL, 0);

    /* Add the NOP actions */
    XtAppAddActions(app_context, actions, XtNumber(actions));

    /* Create edit dialog */
    editBox = create_editbox_widget("edit_box", toplevel);

    /* Initialize the interface */
    mainW = create_mainWindow_widget("main", toplevel);

    /* Create a default label string for the clipboard */
    generic_label = XmStringCreateLocalized("Generic Label");

    /* Register the drawing area as a drop site */
    XmeDropSink(target, NULL, 0);

    XtManageChild(mainW);

    /* Install our special error handler */
    XSetErrorHandler((XErrorHandler) ErrorHandler);

    /* Add the UTM callbacks on the drawing area */

    XtAddCallback(target, XmNdestinationCallback, 
		  (XtCallbackProc) targetDestinationCallback, NULL);
    XtAddCallback(target, XmNconvertCallback, 
		  (XtCallbackProc) targetConvertCallback, NULL);

    /* Add other callbacks on the drawing area */
    XtAddCallback(target, XmNinputCallback,
		  (XtCallbackProc) inputCallback, NULL);
    
    XtRealizeWidget(toplevel);

    XtAppMainLoop(app_context);
}

/***************************************************
 * function setMode				  
 * Set the mode to PRIMARY, SECONDARY or CLIPBOARD
 ***************************************************/
void 
setMode(Widget widget, int param, XmAnyCallbackStruct *callback_data)
{
  Tmode = param;
}

void 
quitCB(Widget widget, char *tag, XmAnyCallbackStruct *callback_data)
{
  exit(0);
}

/******************************************************************
 * cut, copy and paste perform the apropriate actions
 * depending on which mode we're in.  They use the widget writer's
 * API for this.  Normally application developers won't need to
 * call these,  but this is a pseudo widget writing case as we're
 * adding interchange behavior to drawing area.  This is an example
 * of where the widget writing API should be used by app. developers
 ******************************************************************/

void 
cut(Widget widget, char *ignore, XmAnyCallbackStruct *callback_data)
{
  if (Tmode == Mprimary) {
    XmePrimarySource(target, 0);
    XmToggleButtonSetState(op, True, False);
  } else if (Tmode == Mnamed) {
    XmeNamedSource(target, XInternAtom(XtDisplay(target), _NAMED, False), 0);
    XmToggleButtonSetState(onamed, True, False);
  } else if (Tmode == Msecondary) {
    XmeSecondarySource(target, 0);
    XmeSecondaryTransfer(target, 
			 XInternAtom(XtDisplay(target), 
				     "COMPOUND_TEXT", False),
			 XmMOVE, 0);
  } else if (Tmode == Mclipboard)
    XmeClipboardSource(target, XmMOVE, 0, generic_label);
}


void 
copy(Widget widget, char *ignore, XmAnyCallbackStruct *callback_data)
{
  if (Tmode == Mprimary) {
    XmePrimarySource(target, 0);
    XmToggleButtonSetState(op, True, False); 
  } else if (Tmode == Mnamed) {
    XmeNamedSource(target, XInternAtom(XtDisplay(target), _NAMED, False), 0);
    XmToggleButtonSetState(onamed, True, False);
  } else if (Tmode == Msecondary) {
    XmeSecondarySource(target, 0);
    XmeSecondaryTransfer(target, 
			 XInternAtom(XtDisplay(target), 
				     "COMPOUND_TEXT", False),
			 XmCOPY, 0);
  } else if (Tmode == Mclipboard)
    XmeClipboardSource(target, XmCOPY, 0, generic_label);
}

void 
paste(Widget widget, char *ignore, XmAnyCallbackStruct *callback_data)
{
  if (Tmode == Mprimary) {
    XmePrimarySink(target, XmCOPY, NULL, 0);
  } else if (Tmode == Mnamed) {
    XmeNamedSink(target, XInternAtom(XtDisplay(target), _NAMED, False),
		 XmCOPY, NULL, 0);
  } else if (Tmode == Msecondary) {
    XmeSecondarySink(target, 0);
    XmToggleButtonSetState(omd, True, False);
  } else if (Tmode == Mclipboard)
    XmeClipboardSink(target, XmCOPY, 0);
}

void 
pasteMove(Widget widget, char *ignore, XmAnyCallbackStruct *callback_data)
{
  if (Tmode == Mprimary) {
    XmePrimarySink(target, XmMOVE, 0, 0);
  } else if (Tmode == Mnamed) {
    XmeNamedSink(target, XInternAtom(XtDisplay(target), _NAMED, False),
		 XmMOVE, NULL, 0);
  } else if (Tmode == Msecondary) {
    XmeSecondarySink(target, 0);
    XmToggleButtonSetState(omd, True, False);
  }
  else if (Tmode == Mclipboard)
    XmeClipboardSink(target, XmMOVE, 0);
}

/* Error handler for X protocol errors.  Display error text in text1,2 */

static int 
ErrorHandler(Display *display, XErrorEvent *event)
{
  char buffer[80];
  char errortext[100];
  XmString tmp;

  XGetErrorText(display, event -> error_code, errortext, 100);

  tmp = XmStringCreateLocalized(errortext);
  XmCSTextSetString(text2, tmp);
  XmStringFree(tmp);

  sprintf(buffer, "XID %d serial %d major %d minor %d",
	  event -> resourceid,
	  event -> request_code,
	  event -> minor_code);

  tmp = XmStringCreateLocalized(buffer);
  XmCSTextSetString(text1, tmp);
  XmStringFree(tmp);

  return 0;
}

void 
inputCallback(Widget w, XtPointer ignore, XmDrawingAreaCallbackStruct *cbstruct)
{
  XEvent *event = cbstruct -> event;

  if (event -> type == ButtonPress &&
      event -> xbutton.button == 2) {
    XmeDragSource(w, NULL, event, NULL, 0);
  }
}

static void 
NOP(Widget widget, XEvent *event, String *params, Cardinal *num_params)
{
  /* We do nothing in the action. */
}

